File: Data Structures/Hash/LV2_42577.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>

// time over

// hash로 구현했으므로 1,000,000 * 20

using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    
    // 번호의 존재 여부를 저장할 map
    unordered_map <string, int> phone_map;
    
    for (auto number : phone_book) {
        phone_map[number] = 1;
    }
    
    // 각 번호를 앞에서부터 parsing하여 다른 번호의 접두어가 있는지 확인
    for (int i = 0; i < phone_book.size(); i++) {
        
        // parsing한 번호를 저장할 문자열
        string number = "";
        for (int j = 0; j < phone_book[i].size(); j++) {
            
            number += phone_book[i][j];
            
            // 전체 map중 접두어가 있는지 hash로 확인, 자기 자신은 제외
            if (phone_map[number] && number != phone_book[i]) {
                return false;
            }
        }
    }
    
    return answer;
}

File: Data Structures/Hash/LV2_42577.py
# N은 최대 1,000,000
# N + N-1 + N-2 이렇게 검사하는 건 불가능
# 접두어를 파싱해서 이를 저장 후 해시로 O(1) 조회해야함
# 1) N 돌며 하나씩 파싱 후 해시로 저장
#       유효하지 않은 방법
# 2) N 돌며 하나씩 온전히 해시로 저장
#    N 다시 돌며 파싱 후 해시 검사
#    => O(N) * 20 이므로 통과 가능
def solution(phone_book):
    answer = True
    hash = {}
    for number in phone_book:
        hash[number] = 1
    for number in phone_book:
        tmp = ''
        for n in number:
            tmp += n
            if tmp in hash and tmp != number:
                return False
    return answer